contracts/exchangeIssuance/ExchangeIssuanceLeveraged.sol

```diff
@@ -0,0 +1,985 @@
+     *
+     * @param _token    Token to approve
+     * @param _spender  Spender address to approve
+     */
+    function _safeApprove(IERC20 _token, address _spender, uint256 _requiredAllowance) internal {
+        uint256 allowance = _token.allowance(address(this), _spender);
+        if (allowance < _requiredAllowance) {
+            _token.safeIncreaseAllowance(_spender, MAX_UINT96 - allowance);
+        }
+    }
+
+
+
+    /**
+     * Redeems a given amount of SetToken.
+     *
+     * @param _setToken     Address of the SetToken to be redeemed
+     * @param _amount       Amount of SetToken to be redeemed
+     */
+    function _redeemExactSet(ISetToken _setToken, uint256 _amount) internal returns (uint256) {
+        _setToken.safeTransferFrom(msg.sender, address(this), _amount);
+        debtIssuanceModule.redeem(_setToken, _amount, address(this));
+    }
+
+
+    /**
+     * Swap exact tokens for another token on a given DEX.
+     *
+     * @param _exchange     The exchange on which to peform the swap
+     * @param _tokenIn      The address of the input token
+     * @param _tokenOut     The address of the output token
+     * @param _amountIn     The amount of input token to be spent
+     *
+     * @return              The amount of output tokens
+     */
+    function _swapExactTokensForTokens(Exchange _exchange, address _tokenIn, address _tokenOut, uint256 _amountIn) internal returns (uint256) {
+        if (_tokenIn == _tokenOut) {
+            return _amountIn;
+        }
+
+        address[] memory path = _generatePath(_tokenIn, _tokenOut);
+        //TODO: Review if we have to set a non-zero minAmountOut
+        return _getRouter(_exchange).swapExactTokensForTokens(_amountIn, 0, path, address(this), block.timestamp)[1];
+    }
+
+    /**
+     * Swap tokens for exact amount of output tokens on a given DEX.
+     *
+     * @param _exchange     The exchange on which to peform the swap
+     * @param _tokenIn      The address of the input token
+     * @param _tokenOut     The address of the output token
+     * @param _amountOut    The amount of output token required
+     * @param _maxAmountIn  Maximum amount of input token to be spent
+     *
+     * @return              The amount of input tokens spent
+     */
+    function _swapTokensForExactTokens(Exchange _exchange, address _tokenIn, address _tokenOut, uint256 _amountOut, uint256 _maxAmountIn) internal returns (uint256) {
+        if (_tokenIn == _tokenOut) {
+            return _amountOut;
+        }
+        address[] memory path = _generatePath(_tokenIn, _tokenOut);
+        uint256 result = _getRouter(_exchange).swapTokensForExactTokens(_amountOut, _maxAmountIn, path, address(this), block.timestamp)[0];
+        return result;
+    }
+
+    function _generatePath(address _tokenIn, address _tokenOut) internal view returns (address[] memory) {
+        address[] memory path;
+        if(_tokenIn == INTERMEDIATE_TOKEN || _tokenOut == INTERMEDIATE_TOKEN){
```

Comment:
The intermediate token is the highest liquidity token on the exchange, which was hard coded to be "WETH" in the original ExchangeIssuance contract. 
Is there a reason why we would handle this differently here than in the original ExchangeIssuance ? 

ImO if we maximize for flexibility by letting the user pass an "intermediate token" for each trade we might as well just let him pass in the whole `path` right ? 

For context: The only reason why I separated "INTERMEDIATE_TOKEN" and "WETH" is that on Polygon the wrapped native token (represented by variable `WETH` but would actually be wmatic) might be different from the most liquid token on any given dex.

Reply:
Yeah I think having the intermediate token be hardcoded globally will be a pain point since liquidity is pretty split between wMATIC and WETH. Being able to pass in the path will be nice. However, you should take a look at how Uniswap V3 paths work before you make this change, since this might change how you think about the best format to pass in the path.

Reply:
Makes sense. I will revisit this issue then, when I add the UniV3 support and resolve it for now